### PR TITLE
Transfer support for Cosmos3 ModularPipeline

### PR DESCRIPTION
**Pull request by yzhautouskay**
# What does this PR do?

Adds the transfer (structure-controlled) path to the **Cosmos3 Modular pipeline** (`Cosmos3OmniModularPipeline` / `Cosmos3OmniBlocks`), letting users condition video generation on one or more control videos (segmentation, depth, edge/canny, blur, world-scenario map) in addition to the text prompt. This mirrors the transfer path in the task-based `Cosmos3OmniPipeline` and is element-wise consistent with it

## What this adds
- A `transfer` workflow for the modular pipeline, triggered automatically when `control_videos` is passed -- no new pipeline class, it coexists with the existing modular modes (`text2image`, `text2video`, `image2video`, `video2video`, action, sound).
- Transfer is implemented as first-class modular blocks (matching how the existing modalities are built), not a monolithic step:
  - text: `Cosmos3TransferTextStep`
  - before-denoise: `Cosmos3TransferSetupStep`, `Cosmos3TransferPrepareLatentsStep`, `Cosmos3TransferPackSequenceStep`, `Cosmos3TransferSetTimestepsStep`
  - denoise: per-chunk loop blocks (`Cosmos3TransferChunkVaeEncoderStep` -> prepare -> denoise -> scheduler), with the denoiser consuming the packed static contexts through the shared `denoiser_input_fields` mechanism
  - decode: `Cosmos3TransferDecodeChunkStep`, `Cosmos3TransferStitchStep`
  - composed/wired in `modular_blocks_cosmos3.py` via the `Auto*` blocks (selected on `control_videos`), with `"transfer"` added to `available_workflows`.

## Before submitting
- [ ] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [ ] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [ ] Did you self-review the diff against [`.ai/review-rules.md`](https://github.com/huggingface/diffusers/blob/main/.ai/review-rules.md)?
- [ ] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [ ] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)? (important for complex PRs)
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?
- [x] Are you the author (or part of the team) of the model/pipeline (only applicable for model/pipeline related PRs)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/modular_pipelines/cosmos/encoders.py [hunk 1]
```diff
@@ -144,6 +144,139 @@ def __call__(self, components: Cosmos3OmniModularPipeline, state: PipelineState)
...
+            num_frames=block_state.chunk_frames,
+            height=block_state.height,
+            width=block_state.width,
+            fps=block_state.fps,
+            use_system_prompt=block_state.use_system_prompt,
+            add_resolution_template=block_state.add_resolution_template,
+            add_duration_template=block_state.add_duration_template,
+            action_mode=None,
+            action_view_point=None,
+            transfer_mode=True,
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/cosmos/encoders.py]: I assume some code got lost when you resolved the merge conflict 

What's  `transfer_mode`? Do you need a special text step for transfer? If so, can you write one here on the block (instead of modifying the one on the standard pipeline)?

File: src/diffusers/modular_pipelines/cosmos/before_denoise.py [hunk 2]
```diff
@@ -960,3 +963,392 @@ def __call__(self, components: Cosmos3OmniModularPipeline, state: PipelineState)
         )
         self.set_block_state(state, block_state)
         return components, state
+
+
+class Cosmos3TransferSetupStep(ModularPipelineBlocks):
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/cosmos/before_denoise.py]: I think we should move this to either encoder.py or before_encoder.py, since it runs before the text step

File: src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py [hunk 3]
```diff
@@ -566,41 +648,137 @@ def outputs(self):
...
+        )
+
+    @property
+    def inputs(self) -> list[InputParam]:
+        return super().inputs + [
+            InputParam(name="num_chunks", type_hint=int, required=True, description="Number of autoregressive chunks.")
+        ]
+
+    @torch.no_grad()
+    def __call__(self, components: Cosmos3OmniModularPipeline, state: PipelineState) -> PipelineState:
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py:825]: Can you consider using the same pattern as Helio for its chunk loop? https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/helios/denoise.py

Basically, we make the outer chunk loop a proper `LoopSequentialPipelineBlocks` (see https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/helios/denoise.py#L706) and write the inner denoiser loop as a plain block (see https://github.com/huggingface/diffusers/blob/main/src/diffusers/modular_pipelines/helios/denoise.py#L355)

**Review by yiyixuxu — COMMENTED** [review]
thanks a lot, i left some feedbacks

File: src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py (hunk 3 above)
Comment (yzhautouskay) [reply src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py:825]: Yeah, I have looked at Helios pipeline, and the constraint was that `LoopSequentialPipelineBlocks` rejects non-leaf sub-blocks, and our inner timestep loop (`Cosmos3TransferDenoiseStep` -> `Cosmos3DenoiseLoopWrapper`) is itself a `LoopSequentialPipelineBlocks`, so the two loops can't both be loop-wrappers

And I went with keeping the inner loop as the decomposed `LoopSequentialPipelineBlocks` because that's the structure every other cosmos modality uses (vision/sound/action all subclass `Cosmos3DenoiseLoopWrapper`), leaving the chunk loop as the only transfer-specific piece, doing it with `__call__` override of `SequentialPipelineBlocks`

Your proposal also works, but the transfer's inner denoise becomes a monolithic leaf block, diverging from the other modalities (cross-pipeline consistency with Helios vs. intra-cosmos consistency)

Let me know if I am missing something, and you would prefer the refactor

Comment (yiyixuxu) [reply src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py:825]: i started a PR on a better loop block design here https://github.com/huggingface/diffusers/pull/14159/changes

so that we can provide a unified loop abstraction that can be freely composed - you can put a loop as sub-block of another loop and to create the nested chunk loop here; so that user don't have to implement custom looop block with a `__call__` override 

I think the custom loop block here does not work with `get_workflow()`, it just flattens all the subblocks into a `SequentialPipelineblocks` . i.e.. i think a workflow-specific pipeline created this way won't work

```py
transfer_pipe= Cosmos3OmniBlocks().get_workflow("transfer").init_pipeline(...)
out = transfer_pipe(...)
```

but my PR was still in WIP and I don't want to block this PR. I think we can merge this first and we can refactor under my PR? i will tag you for review once it's ready

Comment (yzhautouskay) [reply src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py:825]: Indeed, `get_workflow("transfer")` is currently broken. I removed transfer from `_workflow_map` for now and added a clear exception explaining the limitation and that transfer remains available through the full pipeline.

It would be great to merge this first and refactor transfer to the new nested-loop abstraction once your PR is ready. Please tag me for the review

**Review by yiyixuxu — APPROVED** [review]
thanks

File: src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py [hunk 4]
```diff
@@ -566,41 +648,135 @@ def outputs(self):
...
+        state.set("output_chunks", [])
+        state.set("previous_output", None)
+        for chunk_id in range(num_chunks):
+            state.set("chunk_id", chunk_id)
+            for _, block in self.sub_blocks.items():
+                components, state = block(components, state)
+        return components, state
+
+
+class Cosmos3TransferCoreDenoiseStep(SequentialPipelineBlocks):
```
Comment (yiyixuxu) [inline src/diffusers/modular_pipelines/cosmos/modular_blocks_cosmos3.py:837]:
```suggestion
# auto_docstring
class Cosmos3TransferCoreDenoiseStep(SequentialPipelineBlocks):
```

can you add auto_docstring for all the assembled blocks you added? and make sure don't have any "TODO"s after you run `python utils/modular_auto_docstring.py --fix_and_overwrite`

**Comment by yiyixuxu** [comment]
@bot /style

**Comment by yiyixuxu** [comment]
@bot/ style
